pkg/services/datasources/service.go

```diff
@@ -70,9 +72,37 @@ func ProvideService(bus bus.Bus, store *sqlstore.SQLStore, secretsService secret
 	s.Bus.AddHandler(s.UpdateDataSource)
 	s.Bus.AddHandler(s.GetDefaultDataSource)
 
+	ac.RegisterAttributeScopeResolver(NewDatasourceNameScopeResolver(store))
+
 	return s
 }
 
+type DataSourceRetriever interface {
+	GetDataSource(ctx context.Context, query *models.GetDataSourceQuery) error
+}
+
+// NewDatasourceNameScopeResolver provides an AttributeScopeResolver able to
+// translate a scope prefixed with "datasources:name:" into an id based scope.
+func NewDatasourceNameScopeResolver(db DataSourceRetriever) (string, accesscontrol.AttributeScopeResolveFunc) {
+	dsNameResolver := func(ctx context.Context, orgID int64, initialScope string) (string, error) {
+		dsNames := strings.Split(initialScope, ":")
+		if dsNames[0] != "datasources" || len(dsNames) != 3 {
```

Comment:
I'll try to explain why it makes sense for me, you might not agree with it and we can change it.

Since this is an exported function I can't be sure who's going to use it and for what. We utilize this func signature as an interface but at no point we explicit that the caller should check the scope provided is composed of 3 parts and is targeted to `"datatsources"`. Taken in isolation this unit will not work unless these conditions are validated and that's why they're checked here.

PS: can also be argued that it's the Resolver's job to verify if the provided scope applies to it instead of the caller but that opens another can of worms :sweat_smile:

Reply:
I see what you're saying. I guess it doesn't hurt to double-check that the resolver conditions are met before resolving the scope. Just in case :)